src/index/ivf/ivf.cc

```diff
@@ -751,6 +811,68 @@ IvfIndexNode<DataType, IndexType>::RangeSearch(const DataSet& dataset, const Con
     return GenResultDataSet(nq, ids, distances, lims);
 }
 
+template <typename DataType, typename IndexType>
+expected<std::vector<std::shared_ptr<IndexNode::iterator>>>
+IvfIndexNode<DataType, IndexType>::AnnIterator(const DataSet& dataset, const Config& cfg,
+                                               const BitsetView& bitset) const {
+    if (!index_) {
+        LOG_KNOWHERE_WARNING_ << "creating iterator on empty index";
+        return expected<std::vector<std::shared_ptr<IndexNode::iterator>>>::Err(Status::empty_index,
+                                                                                "index not loaded");
+    }
+    if (!index_->is_trained) {
+        LOG_KNOWHERE_WARNING_ << "index not trained";
+        return expected<std::vector<std::shared_ptr<IndexNode::iterator>>>::Err(Status::index_not_trained,
+                                                                                "index not trained");
+    }
+    // only support IVFFlat and IVFFlatCC;
+    if constexpr (!std::is_same<faiss::IndexIVFFlatCC, IndexType>::value &&
+                  !std::is_same<faiss::IndexIVFFlat, IndexType>::value) {
+        LOG_KNOWHERE_WARNING_ << "Current index_type: " << Type() << ", only IVFFlat and IVFFlatCC support Iterator.";
+        return expected<std::vector<std::shared_ptr<IndexNode::iterator>>>::Err(Status::not_implemented,
+                                                                                "index not supported");
+    } else {
+        auto dim = dataset.GetDim();
+        auto rows = dataset.GetRows();
+        auto data = dataset.GetTensor();
+
+        auto vec = std::vector<std::shared_ptr<IndexNode::iterator>>(rows, nullptr);
+
+        const IvfConfig& ivf_cfg = static_cast<const IvfConfig&>(cfg);
+        bool is_cosine = IsMetricType(ivf_cfg.metric_type.value(), knowhere::metric::COSINE);
+
+        size_t nprobe = ivf_cfg.nprobe.value();
+
+        try {
+            std::vector<folly::Future<folly::Unit>> futs;
+            futs.reserve(rows);
+            for (int i = 0; i < rows; ++i) {
+                futs.emplace_back(search_pool_->push([&, index = i] {
+                    auto cur_query = (const float*)data + index * dim;
+                    std::unique_ptr<float[]> copied_norm_query = nullptr;
+                    if (is_cosine) {
+                        copied_norm_query = CopyAndNormalizeVecs(cur_query, 1, dim);
+                    }
+
+                    // the iterator only own the copied_norm_query.
+                    vec[index].reset(
+                        new iterator(index_.get(), cur_query, std::move(copied_norm_query), bitset, nprobe));
```

Comment:
this implies that if the metric is COSINE, then the `vec` will duplicate the amount of needed data in memory (because of `copied_norm_query`). I'm mentioning just in case :) Technically, one could do lazy evaluation

Reply:
Currently, the iterator invokes the `UpdateNext()` directly within the constructor, allowing the Next() function to be pre-calculated. It appears that there is no room for lazy evaluation. Do you have any better suggestions?

Reply:
Let's just launch as is for now